### PR TITLE
 Add Export Safetensors format  Support for Qwen3_5 Model

### PR DESCRIPTION
**Pull request by laxmareddyp**
## Description of the change
 
Numerics verification screenshots:

Screenshot 2026-04-29 at 12 10 28 AM
Screenshot 2026-04-29 at 12 11 07 AM
Screenshot 2026-04-29 at 12 11 41 AM

### CODE REVIEW

File: keras_hub/src/utils/transformers/export/qwen3_5.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+  - get_qwen3_5_config: Backbone config → HF config dict
+  - get_qwen3_5_weights_map: Backbone weights → HF weight name mapping
+  - get_qwen3_5_tokenizer_config: Tokenizer → HF tokenizer config dict
+"""
+
+import keras.ops as ops
+
+
+def get_qwen3_5_config(backbone):
+    """Convert KerasHub Qwen3.5 backbone config to HF config dictionary."""
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5.py]: Still other export model doesn't follow this pattern but still i have updated the code.

File: keras_hub/src/utils/transformers/export/qwen3_5.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+    return hf_config
+
+
+def get_qwen3_5_weights_map(backbone, include_lm_head=False):
+    """Create a weights map from KerasHub Qwen3.5 backbone to HF format.
+
+    This is the inverse of the import converter in convert_qwen3_5.py.
+    Each HF weight key maps to the corresponding Keras tensor, with
+    appropriate transposes to reverse the import hook_fns.
+    """
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5.py:112]: Still other export model doesn't follow this pattern but still i have updated the code.

File: keras_hub/src/utils/transformers/export/qwen3_5.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+        weights_map[f"{merger_prefix}.linear_fc2.weight"] = ops.transpose(
+            vis.merger.fc2.kernel
+        )
+        weights_map[f"{merger_prefix}.linear_fc2.bias"] = vis.merger.fc2.bias
+
+    return weights_map
+
+
+def get_qwen3_5_tokenizer_config(tokenizer):
+    """Convert KerasHub Qwen3.5 tokenizer config to HF format."""
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5.py]: Still other export model doesn't follow this pattern but still i have updated the code.

File: keras_hub/src/utils/transformers/export/qwen3_5.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        if layer_type == "full_attention":
+            attn = decoder_layer._self_attention_layer
+
+            # Q projection: Keras shape (hidden_dim, num_heads, head_dim*2)
+            # → HF shape (num_heads * head_dim * 2, hidden_dim), i.e. transpose
+            q_kernel = attn._query_dense.kernel
+            q_kernel = ops.reshape(q_kernel, (backbone.hidden_dim, -1))
+            weights_map[f"{prefix}.self_attn.q_proj.weight"] = ops.transpose(
+                q_kernel
+            )
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5.py:139]: These are EinsumDense layers with 3D kernels (For example. shape (hidden_dim, num_heads, head_dim*2)), not standard 2D Dense layers. The reshape is necessary to flatten (num_heads, head_dim) into a single dim before transposing to HF's 2D format.

File: keras_hub/src/utils/transformers/export/qwen3_5.py [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+                    vis.in_channels,
+                )
+            )
+        if not vis.pos_embed.built:
+            vis.pos_embed.build((None,))
+        for blk in vis.blocks:
+            if not blk.built:
+                blk.build((None, vis.hidden_size))
+        if not vis.merger.built:
+            vis.merger.build((None, vis.hidden_size))
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5.py]: These build calls are necessary — the Qwen3.5 vision encoder sublayers use `lazy building` and aren't automatically built when loaded from a preset. The same pattern exists in the official import converter at convert_qwen3_5.py  https://github.com/keras-team/keras-hub/blob/8536f4da17e5af2660226bf6c0e647952bbf8171/keras_hub/src/utils/transformers/convert_qwen3_5.py#L362 Without these calls, accessing vision weights raises AttributeError.

File: keras_hub/src/utils/transformers/export/qwen3_5.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5.py:139]: Acknowledged. Keeping the current reshape and transpose pattern for consistency with the other exporters and the inverse import converter in convert_qwen3_5.py.

File: keras_hub/src/utils/transformers/export/qwen3_5_test.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        export_path = os.path.join(temp_dir, "export_task")
+        export_to_safetensors(keras_model, export_path)
+
+        # Patch config for our tiny vocab
+        config_path = os.path.join(export_path, "config.json")
+        with open(config_path, "r") as f:
+            cfg = json.load(f)
+        cfg["eos_token_id"] = 2  # <|im_end|>
+        with open(config_path, "w") as f:
+            json.dump(cfg, f, indent=2)
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5_test.py]: `eos_token_id` is a tokenizer attribute, not a backbone config field. `get_qwen3_5_config()` only receives the backbone. The token ID is already exported correctly in `tokenizer_config.json` via eos_token: '<|im_end|>'. The test patch is needed only because the unit test uses a tiny custom vocab where HF can't resolve the token automatically.

File: keras_hub/src/utils/transformers/export/qwen3_5_test.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+            "Linear num key heads do not match",
+        )
+        self.assertEqual(
+            text_cfg.linear_num_value_heads,
+            backbone.linear_num_value_heads,
+            "Linear num value heads do not match",
+        )
+
+        # 8. Compare Logits (round-trip numerical parity)
+        import torch
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5_test.py]: Done. Moved.

[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5_test.py]: Updated in previous comment

File: keras_hub/src/utils/transformers/export/qwen3_5.py [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+        "attn_output_gate": True,
+        "tie_word_embeddings": backbone.tie_word_embeddings,
+        "rope_parameters": rope_parameters,
+        "layer_types": backbone.layer_types,
+        "linear_num_key_heads": backbone.linear_num_key_heads,
+        "linear_num_value_heads": backbone.linear_num_value_heads,
+        "linear_key_head_dim": backbone.linear_key_head_dim,
+        "linear_value_head_dim": backbone.linear_value_head_dim,
+        "linear_conv_kernel_dim": backbone.linear_conv_kernel_dim,
+        "mamba_ssm_dtype": "float32",
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5.py:48]: This matches the original HF config which hardcodes `mamba_ssm_dtype: 'float32'.` It's a model architecture constant for numerical stability in the SSM recurrence, not a compute dtype setting. Using dtype_policy.compute_dtype would incorrectly export 'bfloat16' or 'float16' when the model is loaded in mixed precision.

File: keras_hub/src/utils/transformers/export/qwen3_5.py [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+
+def get_qwen3_5_tokenizer_config(tokenizer):
+    """Convert KerasHub Qwen3.5 tokenizer config to HF format."""
+    return {
+        "tokenizer_class": "Qwen2Tokenizer",
+        "bos_token": None,
+        "eos_token": "<|im_end|>",
+        "pad_token": "<|endoftext|>",
+        "unk_token": None,
+        "model_max_length": 131072,
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5.py]: model_max_length is a tokenizer constant (max context window), not derived from the model's sequence_length (which is a preprocessor truncation setting, typically much smaller). Fixed the hardcoded value from 131072 to 262144 to match the original HF tokenizer.

File: keras_hub/src/utils/transformers/export/qwen3_5.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5.py]: Updated in previous comment

File: keras_hub/src/utils/transformers/export/qwen3_5.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5.py:112]: Updated in previous comment

File: keras_hub/src/utils/transformers/export/qwen3_5.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5.py]: Updated in previous comment

File: keras_hub/src/utils/transformers/export/qwen3_5.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5.py:48]: Verified `mamba_ssm_dtype: 'float32'` is present in the original HF Qwen3.5 config . While Qwen3.5 uses GatedDeltaNet rather than Mamba, the HF `Qwen3_5TextConfig` class includes this field as the GatedDeltaNet implementation shares SSM infrastructure.

File: keras_hub/src/utils/transformers/export/qwen3_5.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5.py]: These build calls are necessary. The `Qwen3.5 vision encoder sublayers use lazy building` and aren't automatically built when loaded from a preset. The same pattern exists in the official import converter at convert_qwen3_5.py  https://github.com/keras-team/keras-hub/blob/8536f4da17e5af2660226bf6c0e647952bbf8171/keras_hub/src/utils/transformers/convert_qwen3_5.py#L362

File: keras_hub/src/utils/transformers/export/qwen3_5.py [hunk 10]
```diff
@@ -0,0 +1,355 @@
+"""Export utilities for Qwen3.5 models to HuggingFace Transformers format.
+
+Provides three functions following the KerasHub export pattern:
+  - get_qwen3_5_config: Backbone config → HF config dict
+  - get_qwen3_5_weights_map: Backbone weights → HF weight name mapping
+  - get_qwen3_5_tokenizer_config: Tokenizer → HF tokenizer config dict
+"""
```
Comment (sachinprasadhs) [inline keras_hub/src/utils/transformers/export/qwen3_5.py]: Remove this, we don't do file annotation.

File: keras_hub/src/utils/transformers/export/qwen3_5.py [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+        "num_attention_heads": backbone.num_query_heads,
+        "num_key_value_heads": backbone.num_key_value_heads,
+        "intermediate_size": backbone.intermediate_dim,
+        "head_dim": backbone.head_dim,
+        "rms_norm_eps": backbone.layer_norm_epsilon,
+        "hidden_act": "silu",
+        "use_cache": True,
+        "attention_bias": False,
+        "attention_dropout": 0.0,
+        "attn_output_gate": True,
```
Comment (sachinprasadhs) [inline keras_hub/src/utils/transformers/export/qwen3_5.py:39]: I believe we are not saving this in our Keras Hub preset configs? If not, load it from preset configs.

File: tools/checkpoint_export/verify_qwen3_5_export.py [hunk 12]
```diff
@@ -0,0 +1,583 @@
...
+    fields = [
+        "vocab_size",
+        "num_hidden_layers",
+        "num_attention_heads",
+        "num_key_value_heads",
+        "hidden_size",
+        "intermediate_size",
+        "head_dim",
+        "tie_word_embeddings",
+    ]
```
Comment (sachinprasadhs) [inline tools/checkpoint_export/verify_qwen3_5_export.py]: We need to compare ot for all the configs, not just few keys.
We can have seperate functions for validate_configs, validate_token_ids, numerics etc.

File: tools/checkpoint_export/verify_qwen3_5_export.py [hunk 13]
```diff
@@ -0,0 +1,583 @@
...
+                            "fps": VIDEO_FPS,
+                        },
+                        {"type": "text", "text": "Describe this video."},
+                    ],
+                }
+            ]
+            try:
+                text_input = processor.apply_chat_template(
+                    messages, tokenize=False, add_generation_prompt=True
+                )
```
Comment (sachinprasadhs) [inline tools/checkpoint_export/verify_qwen3_5_export.py]: May be we can avoid apply_chat_template and do it in a reguar way of constructing the prompt. It would be easier for the users also to use the models in ceratin way.

**Review by sachinprasadhs — COMMENTED** [review]
Left few comments. I still believe Gemini comments are valid for qwen3_5.py file.

File: keras_hub/src/utils/transformers/export/qwen3_5.py (hunk 10 above)
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5.py]: Updated.

File: keras_hub/src/utils/transformers/export/qwen3_5.py (hunk 11 above)
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5.py:39]: These are architecture constants (not configurable) — the Qwen3.5 backbone doesn't store them as attributes. Same approach as the existing Qwen exporter , which hardcodes hidden_act: 'silu', use_cache: True. All Qwen3.5 variants share these values.

File: tools/checkpoint_export/verify_qwen3_5_export.py (hunk 12 above)
Comment (laxmareddyp) [reply tools/checkpoint_export/verify_qwen3_5_export.py]: Will take care of these.

File: tools/checkpoint_export/verify_qwen3_5_export.py (hunk 13 above)
Comment (laxmareddyp) [reply tools/checkpoint_export/verify_qwen3_5_export.py]: I think `apply_chat_template`  is still valid here, We are loading and testing with the torch weights and trying to find the comparison of the output. We are not using any of the keras-hub weights here for comparison.

File: keras_hub/src/utils/transformers/export/qwen3_5_test.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/utils/transformers/export/qwen3_5_test.py]: Added `eos_token_id: 248044` to get_qwen3_5_config(). The test still patches it because the tiny test vocab (size 10) can't use the production token ID — this is the same pattern used by GPT2's export test

**Review by divyashreepathihalli — APPROVED** [review]

**Comment by laxmareddyp** [comment]
Updating the test suite to align with the recently merged BytePairTokenizer pull request.
